# CSI5155 Machine Learning Assignment 1

## Installing Prerequisites and Dependencies

In [1]:
%pip install pandas
%pip install numpy
%pip install scikit-learn
%pip install joblib

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd;
import numpy as np;
from models import Models;
import constants;

## Load the Dataset

In [3]:
df_all = pd.read_csv('drug_consumption.data', sep=',', header=None, names=constants.column_names_orig);
df_all

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore  ...  ecstasy  heroin  ketamine legalh  lsd meth  \
0    -0.58331 -0.91699  ...      CL0     CL0       CL0    CL0  CL0  CL0   
1     1.43533  0.76096  ...      CL4     CL0       CL2    CL0  CL2  CL3   
2    -0.84732 -1.62090  ...      CL0     CL0       CL0    CL0  CL0  CL0   
3    -0.01928  0.59042  ...      CL0     CL0       CL2    CL0  CL0  CL0   
4    -0.45174 -0.30172  ...      CL1     CL0       CL0    CL1  CL0  CL0   
...       ...      ...  ...      ...     ...       ...    ...  ...  ...   
1880  1.88511  0.76096  ...      CL0     CL0       CL0    CL3  CL3  CL0   
1881  0.58331  0.76096  ...      CL2     CL0       CL0    CL3  CL5  CL4   
1882 -1.27553 -1.77200  ...      CL4     CL0       CL2    CL0  CL2  CL0   
1883  0.29338 -1.62090  ...      CL3     CL0       CL0    CL3  CL3  CL0   
1884  1.65653  1.11406  ...      CL3     CL0       CL0    CL3  CL3  CL0   

     mushrooms nicotine semer  vsa  
0          CL0      CL2   CL0  CL0  
1          CL0      CL4   CL0  CL0  
2          CL1      CL0   CL0  CL0  
3          CL0      CL2   CL0  CL0  
4          CL2      CL2   CL0  CL0  
...        ...      ...   ...  ...  
1880       CL0      CL0   CL0  CL5  
1881       CL4      CL5   CL0  CL0  
1882       CL2      CL6   CL0  CL0  
1883       CL3      CL4   CL0  CL0  
1884       CL3      CL6   CL0  CL2  

[1885 rows x 32 columns]

Observations:
- The original dataset consists of 1885 instances. 
- Each of which has 32 features in total. 

## Handling Data in the dataset to match with our problem

### Group and Convert multiple categories into 2 unique categories (for the binary classification)

In [4]:
num_types = ['int64', 'float64'];
for i in df_all.columns:
    if (df_all[i].dtype not in num_types):
        df_all[i] = df_all[i].apply(lambda x: "non-user" if int(x.replace("CL", "")) <= 1 else "user");
df_all

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore  ...   ecstasy    heroin  ketamine    legalh       lsd  \
0    -0.58331 -0.91699  ...  non-user  non-user  non-user  non-user  non-user   
1     1.43533  0.76096  ...      user  non-user      user  non-user      user   
2    -0.84732 -1.62090  ...  non-user  non-user  non-user  non-user  non-user   
3    -0.01928  0.59042  ...  non-user  non-user      user  non-user  non-user   
4    -0.45174 -0.30172  ...  non-user  non-user  non-user  non-user  non-user   
...       ...      ...  ...       ...       ...       ...       ...       ...   
1880  1.88511  0.76096  ...  non-user  non-user  non-user      user      user   
1881  0.58331  0.76096  ...      user  non-user  non-user      user      user   
1882 -1.27553 -1.77200  ...      user  non-user      user  non-user      user   
1883  0.29338 -1.62090  ...      user  non-user  non-user      user      user   
1884  1.65653  1.11406  ...      user  non-user  non-user      user      user   

          meth mushrooms  nicotine     semer       vsa  
0     non-user  non-user      user  non-user  non-user  
1         user  non-user      user  non-user  non-user  
2     non-user  non-user  non-user  non-user  non-user  
3     non-user  non-user      user  non-user  non-user  
4     non-user      user      user  non-user  non-user  
...        ...       ...       ...       ...       ...  
1880  non-user  non-user  non-user  non-user      user  
1881      user      user      user  non-user  non-user  
1882  non-user      user      user  non-user  non-user  
1883  non-user      user      user  non-user  non-user  
1884  non-user      user      user  non-user      user  

[1885 rows x 32 columns]

Observation:
- All categorical fields (which represents whether the person has taken that drug before) has been replaced with values suitable for  a binary classfication problem.

In [5]:
# save the modified DataFrame into a new .csv file
df_all.to_csv('drug_consumption.modified', index=False)

### Construct 2 datasets

#### People who consume Chocolate

In [6]:
# find a list of column names in consideration
cols = [];
for i in df_all.columns:
    if (df_all[i].dtype in num_types):
        cols.append(i);
cols.append('choc');
cols

['ID',
 'age',
 'gender',
 'education',
 'country',
 'ethnicity',
 'nscore',
 'escore',
 'oscore',
 'ascore',
 'cscore',
 'impuslive',
 'ss',
 'choc']

In [7]:
df_dataset_choco = df_all[cols];
df_dataset_choco

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore   cscore  impuslive       ss  choc  
0    -0.58331 -0.91699 -0.00665   -0.21712 -1.18084  user  
1     1.43533  0.76096 -0.14277   -0.71126 -0.21575  user  
2    -0.84732 -1.62090 -1.01450   -1.37983  0.40148  user  
3    -0.01928  0.59042  0.58489   -1.37983 -1.18084  user  
4    -0.45174 -0.30172  1.30612   -0.21712 -0.21575  user  
...       ...      ...      ...        ...      ...   ...  
1880  1.88511  0.76096 -1.13788    0.88113  1.92173  user  
1881  0.58331  0.76096 -1.51840    0.88113  0.76540  user  
1882 -1.27553 -1.77200 -1.38502    0.52975 -0.52593  user  
1883  0.29338 -1.62090 -2.57309    1.29221  1.22470  user  
1884  1.65653  1.11406  0.41594    0.88113  1.22470  user  

[1885 rows x 14 columns]

Observations:
- We have 14 features in consideration in the first dataset.
- Meanwhile, the number of instances remains unchanged. 

#### People who consume Magic Mushrooms

In [8]:
# find a list of column names in consideration
cols = [];
for i in df_all.columns:
    if (df_all[i].dtype in num_types):
        cols.append(i);
cols.append('mushrooms');
cols

['ID',
 'age',
 'gender',
 'education',
 'country',
 'ethnicity',
 'nscore',
 'escore',
 'oscore',
 'ascore',
 'cscore',
 'impuslive',
 'ss',
 'mushrooms']

In [9]:
df_dataset_mushrooms = df_all[cols];
df_dataset_mushrooms

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore   cscore  impuslive       ss mushrooms  
0    -0.58331 -0.91699 -0.00665   -0.21712 -1.18084  non-user  
1     1.43533  0.76096 -0.14277   -0.71126 -0.21575  non-user  
2    -0.84732 -1.62090 -1.01450   -1.37983  0.40148  non-user  
3    -0.01928  0.59042  0.58489   -1.37983 -1.18084  non-user  
4    -0.45174 -0.30172  1.30612   -0.21712 -0.21575      user  
...       ...      ...      ...        ...      ...       ...  
1880  1.88511  0.76096 -1.13788    0.88113  1.92173  non-user  
1881  0.58331  0.76096 -1.51840    0.88113  0.76540      user  
1882 -1.27553 -1.77200 -1.38502    0.52975 -0.52593      user  
1883  0.29338 -1.62090 -2.57309    1.29221  1.22470      user  
1884  1.65653  1.11406  0.41594    0.88113  1.22470      user  

[1885 rows x 14 columns]

Observations:
- We have 14 features in consideration in the second dataset.
- Meanwhile, the number of instances also remains unchanged. 

## A. Machine Learning - Supervised Learning

### Load the Models

In [12]:
models = Models();
resModels = models.getModels(
    constants.descisionTree, # Single Decision Tree classifier
    constants.randForest, # Random Forest classifier
    constants.svm, # Support Vector Machine (SVM) - SVC with RBF
);
print(resModels);
# save all models
models.saveModels();

SyntaxError: invalid syntax. Perhaps you forgot a comma? (1939862152.py, line 4)

### Classification task for Chocolate users